README.md
# racir

[![Build Status](https://travis-ci.com/jstinzi/racir.svg?branch=master)](https://travis-ci.com/jstinzi/racir)

[![DOI](https://zenodo.org/badge/149156152.svg)](https://zenodo.org/badge/latestdoi/149156152)

The goal of racir is to provide fast and easy analysis of rapid A/Ci response
(RACiR TM) data obtained using the LI-COR 6800 portable photosynthesis machine
(LI-COR Biosciences, Lincoln, NB). 

RACiR is a trademark of LI-COR Biosciences, and used with permission.

Please report bugs or suggest features to Dr. Joseph Ronald Stinziano at
josephstinziano@gmail.com.

##Contents
*read_6800() reads in a tab-delimited LI-COR 6800 file.
*racircalcheck() allows the user to run QAQC on the calibration RACiR data.
*racircal() calibrates one RACiR leaf data frame based on a calibration data
    frame.
*racircalbatch() calibrates multiple RACiR leaf data frames based on a single
    calibration data frame.
    
##Notes & Help
For use on rapid A/Ci response data from other instruments, simply modify the
varnames variable. Please contact Dr. Joseph R. Stinziano for more information.

## Installation

To install the development version, use the following commands. Windows users
must have Rtools installed.

library(devtools)
install_github("jstinzi/racir")


NEWS.md
# racir 2.0.0

- Removed file writing components of functions for CRAN-compatibility
- Added MIT license
- Generalized functions to better work with data from multiple sources
- Increased user-friendliness of functions
- Improved read function
- Updated tests
- Note version 2.0.0 is NOT back-compatible with version 1.0.0

# racir 1.0.0

## Major Changes

- Created the package


tests/testthat.R
library(testthat)
library(racir)

test_check("racir")


tests/testthat/testracircal.R
library(testthat)
library(racir)
context("Calibrating racir files")

set.seed(5)
n <- 50
df1 <- data.frame(obs   = c(runif(n, 1, 200)),
                 E     = c(runif(n, 10 / 10000, 20 / 1000)),
                 A     = c(runif(n, -1, 30)),
                 Ci    = c(runif(n, 50, 700)),
                 gtc   = c(runif(n, 0.02, 0.3)),
                 Ca    = c(runif(n, 300, 900)),
                 CO2_r = c(runif(n, 300, 900)),
                 CO2_s = c(runif(n, 300, 900)),
                 H2O_r = c(runif(n, 10, 20)),
                 H2O_s = c(runif(n, 10, 20)),
                 Tleaf = c(rep(25, n)),
                 Qin   = c(rep(1500, n))
)

set.seed(6)
n <- 50
df2 <- data.frame(obs   = c(runif(n, 1, 200)),
                  E     = c(runif(n, 10 / 10000, 20 / 1000)),
                  A     = c(runif(n, -1, 30)),
                  Ci    = c(runif(n, 50, 700)),
                  gtc   = c(runif(n, 0.02, 0.3)),
                  Ca    = c(runif(n, 300, 900)),
                  CO2_r = c(runif(n, 300, 900)),
                  CO2_s = c(runif(n, 300, 900)),
                  H2O_r = c(runif(n, 10, 20)),
                  H2O_s = c(runif(n, 10, 20)),
                  Tleaf = c(rep(25, n)),
                  Qin   = c(rep(1500, n))
)

data <- list(df1, df2)

testdata <- racircal(caldata = data[[1]], data = data[[2]],
         mincut = 300, maxcut = 900)

test_that("Outputs", {
  expect_equal(object = length(testdata), length(data[[2]]) + 2)
  expect_is(object = testdata, class = "data.frame")
})

testdata <- racircal_advanced(caldata = data[[1]], data = data[[2]],
                     mincut = 300, maxcut = 900, digits = -3)

test_that("Outputs", {
  expect_equal(object = length(testdata), length(data[[2]]) + 3)
  expect_is(object = testdata, class = "data.frame")
})


tests/testthat/testracircalbatch.R
library(testthat)
library(racir)
context("Batch racir analysis")

data <- vector("list", 2)

set.seed(5)
n <- 50
data[[1]] <- data.frame(obs   = c(runif(n, 1, 200)),
                  E     = c(runif(n, 10 / 10000, 20 / 1000)),
                  A     = c(runif(n, -1, 30)),
                  Ci    = c(runif(n, 50, 700)),
                  gtc   = c(runif(n, 0.02, 0.3)),
                  Ca    = c(runif(n, 300, 900)),
                  CO2_r = c(runif(n, 300, 900)),
                  CO2_s = c(runif(n, 300, 900)),
                  H2O_r = c(runif(n, 10, 20)),
                  H2O_s = c(runif(n, 10, 20)),
                  Tleaf = c(rep(25, n)),
                  Qin   = c(rep(1500, n))
)

set.seed(6)
data[[2]] <- data.frame(obs   = c(runif(n, 1, 200)),
                  E     = c(runif(n, 10 / 10000, 20 / 1000)),
                  A     = c(runif(n, -1, 30)),
                  Ci    = c(runif(n, 50, 700)),
                  gtc   = c(runif(n, 0.02, 0.3)),
                  Ca    = c(runif(n, 300, 900)),
                  CO2_r = c(runif(n, 300, 900)),
                  CO2_s = c(runif(n, 300, 900)),
                  H2O_r = c(runif(n, 10, 20)),
                  H2O_s = c(runif(n, 10, 20)),
                  Tleaf = c(rep(25, n)),
                  Qin   = c(rep(1500, n))
)

set.seed(7)
cal <- data.frame(obs   = c(runif(n, 1, 200)),
                        E     = c(runif(n, 10 / 10000, 20 / 1000)),
                        A     = c(runif(n, -1, 30)),
                        Ci    = c(runif(n, 50, 700)),
                        gtc   = c(runif(n, 0.02, 0.3)),
                        Ca    = c(runif(n, 300, 900)),
                        CO2_r = c(runif(n, 300, 900)),
                        CO2_s = c(runif(n, 300, 900)),
                        H2O_r = c(runif(n, 10, 20)),
                        H2O_s = c(runif(n, 10, 20)),
                        Tleaf = c(rep(25, n)),
                        Qin   = c(rep(1500, n))
)

output <- racircalbatch(caldata = cal,
              data = data,
              mincut = 300,
              maxcut = 900)

test_that("Output", {
  expect_is(object = output, class = "list")
  expect_is(object = output[[1]], class = "data.frame")
  expect_is(object = output[[2]], class = "data.frame")
  expect_length(object = output, 2)
  expect_equal(object = length(output[[1]]), length(data[[1]]) + 3)
  expect_equal(object = length(output[[2]]), length(data[[2]]) + 3)
})

output <- racircalbatch_advanced(caldata = cal,
                        data = data,
                        mincut = 300,
                        maxcut = 900,
                        digits = -3)

test_that("Output", {
  expect_is(object = output, class = "list")
  expect_is(object = output[[1]], class = "data.frame")
  expect_is(object = output[[2]], class = "data.frame")
  expect_length(object = output, 2)
  expect_equal(object = length(output[[1]]), length(data[[1]]) + 4)
  expect_equal(object = length(output[[2]]), length(data[[2]]) + 4)
})


R/racircal.R
#' Corrects rapid A/Ci response (RACiR) data from leaves using empty chamber
#' data.
#'
#' \code{racircal} Corrects your RACiR data based on calibration data. Produces
#' corrected A vs. Ci graph. Output is a data frame with corrected RACiR data
#' using variable names Acor and Cicor for the corrected A and Ci values.
#' @param data Data frame with the RACiR response data
#' @param caldata Data frame with the calibration data
#' @param mincut Minimum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the initial chamber mixing. Default value is set to the
#' minimum COR_r value.
#' @param maxcut Maximum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the end of the response. Not needed in all cases. Default
#' value is set to the maximum COR_r value.
#' @param title Title of output graph - useful for batch RACiR corrections.
#' @param varnames Variable names - this allows for the use of this code with
#' other machines and setups where variable names may differ.
#'
#' @return racircal returns a data frame with corrected RACiR data
#' @importFrom stats BIC
#' @importFrom stats lm
#' @importFrom stats predict
#' @importFrom graphics legend
#' @importFrom graphics lines
#' @importFrom graphics plot
#' @export
#' @examples \donttest{
#' #Read in data
#' data <- read_6800(system.file("extdata", "poplar_2", package = "racir"))
#' caldata <- read_6800(system.file("extdata", "cal", package = "racir"))
#' #Correct data
#' data_corrected <- racircal(data = data, caldata = caldata,
#'                            mincut = 350, maxcut = 780, title = "Test")
#' }
#'
racircal <- function(data, caldata, mincut, maxcut, title,
                     varnames = list(A = "A",
                                     Ca = "Ca",
                                     CO2_r = "CO2_r",
                                     E = "E",
                                     gtc = "gtc")){
  #assign variable names
  data$A <- data[, varnames$A]
  data$Ca <- data[, varnames$Ca]
  data$CO2_r <- data[, varnames$CO2_r]
  data$E <- data[, varnames$E]
  data$gtc <- data[, varnames$gtc]

  # Check for title
  title <- ifelse(missing(title) == TRUE, NA, title)
  # Assign cutoffs ------------------------------------------
  mincut <- ifelse(missing(mincut) == TRUE, min(caldata$CO2_r), mincut)
  maxcut <- ifelse(missing(maxcut) == TRUE, max(caldata$CO2_r), maxcut)
  caldata <- caldata[caldata$CO2_r > mincut, ]
  caldata <- caldata[caldata$CO2_r < maxcut, ]

  # Fit polynomials to calibration curve --------------------
  cal1st <- lm(A ~ CO2_r, data = caldata)
  cal2nd <- lm(A ~ poly(CO2_r, 2), data = caldata)
  cal3rd <- lm(A ~ poly(CO2_r, 3), data = caldata)
  cal4th <- lm(A ~ poly(CO2_r, 4), data = caldata)
  cal5th <- lm(A ~ poly(CO2_r, 5), data = caldata)

  # Use BIC to assess best polynomial -----------------------
  bics <- BIC(cal1st, cal2nd, cal3rd, cal4th, cal5th)
  best <- noquote(rownames(bics)[bics$BIC == min(bics$BIC)])

  # Assigns maximum and minimum CO2_r values based on -------
  # calibration data ----------------------------------------
  maxcal <- max(caldata$CO2_r)
  mincal <- min(caldata$CO2_r)
  # Restrict data to calibration range ----------------------
  data <- data[data$CO2_r > mincal, ]
  data <- data[data$CO2_r < maxcal, ]

  # Correct leaf data ---------------------------------------
  ifelse(best == "cal5th", data$Acor <- data$A - predict(cal5th, data),
         ifelse(best == "cal4th", data$Acor <- data$A - predict(cal4th, data),
         ifelse(best == "cal3rd", data$Acor <- data$A - predict(cal3rd, data),
         ifelse(best == "cal2nd", data$Acor <- data$A - predict(cal2nd, data),
            data$Acor <- data$A - predict(cal1st, data)))))
  data$Cicor <- ( ( (data$gtc - data$E / 2) * data$Ca - data$Acor) /
                  (data$gtc + data$E / 2))
  plot(Acor ~ Cicor, data = data, main = title)
  # Remove columns filled with NA ---------------------------
  output <- data[, unlist(lapply(data, function(x) !all(is.na(x))))]
  # Return data frame
  return(output)
}


R/read_6800.R
#' Reads files from the Li-Cor 6800
#'
#' \code{read_6800} Reads Li-Cor 6800 files, which are delimited by spaces and tabs.
#'
#' @param x A Li-Cor 6800 data file name of the form: "mydata".
#'
#' @return read_6800 imports a Li-Cor 6800 file as a data frame
#' @importFrom utils read.delim
#' @export
#'
#'
#'
read_6800 <- function(x) {
  #Read in header information
  header <- read.delim(file = x, header = TRUE, sep = "\t",
                       skip = grep(pattern = "\\[Data\\]",
                                   perl = TRUE,
                                   useBytes = TRUE,
                                   x = readLines(x),
                                   value = FALSE),
                       nrows = 1)
  #Read in data information
  data <- read.delim(file = x, header = FALSE, sep = "\t",
                     skip = grep(pattern = "\\[Data\\]",
                                 perl = TRUE,
                                 useBytes = TRUE,
                                 x = readLines(x),
                                 value = FALSE) + 3)
  #Add header to data
  colnames(data) <- header[1,]
  #Return data
  return(data)
}


R/racircalbatch.R
#' Corrects a batch of rapid A/Ci response (RACiR) data
#'
#' \code{racircalbatch} Corrects your RACiR data files based on a calibration file.
#' Produces diagnostic graphs of A vs. Ci for quality control. Output
#' includes a list of data frames with corrected data.
#'
#' @param data List of data frames with the RACiR response data
#' @param caldata Data frame with the calibration data
#' @param mincut Minimum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the initial chamber mixing. Default value is set to the
#' minimum COR_r value.
#' @param maxcut Maximum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the end of the response. Not needed in all cases. Default
#' value is set to the maximum COR_r value.
#' @param title Vector for titles of output graph - useful for batch RACiR
#' corrections. Length must be equal to data list length
#' @param varnames Variable names - this allows for the use of this code with
#' other machines and setups where variable names may differ.
#'
#' @return racircalbatch calibrates a batch of RACiR data
#' @importFrom stats BIC
#' @importFrom stats lm
#' @importFrom stats predict
#' @importFrom graphics legend
#' @importFrom graphics lines
#' @importFrom graphics plot
#' @export
#'@examples \donttest{
#'#Create a list of files
#'files <- c(system.file("extdata", "poplar_1", package = "racir"),
#'           system.file("extdata", "poplar_2", package = "racir"))
#'data <- vector("list", length(files))
#'for(i in seq_along(files)){
#'  data[[i]] <- read_6800(files[i])
#'  names(data)[i] <- files[i]
#'}
#'
#'caldata <- read_6800(system.file("extdata", "cal", package = "racir"))
#'output <- racircalbatch(caldata = caldata, data = data,
#'                        mincut = 300, maxcut = 780, title = files)
#'}
#'
racircalbatch <- function(caldata, data, mincut, maxcut,
                          title, varnames = list(A = "A",
                                                 Ca = "Ca",
                                                 CO2_r = "CO2_r",
                                                 E = "E",
                                                 gtc = "gtc")){

  # Check for title
  title <- ifelse(missing(title) == TRUE, rep(NA, length(data)), title)
  output <- vector("list", length(data))
  for (i in seq_along(data)){
    output[[i]] <- racircal(caldata = caldata,
                            data = data[[i]],
                            mincut = mincut,
                            maxcut = maxcut,
                            title = title[i],
                            varnames = varnames)
    output[[i]]$ID <- title[i]
  }
  return(output)
  }


R/racircalcheck.R
#' Allows visual checking of rapid A/Ci response (RACiR) calibration data using
#' empty chamber data.
#'
#' \code{racircalcheck} Used to check range of calibration file. Produces
#' diagnostic graphs of A vs. Ci for quality control. Output includes plots
#' for checking and confirming cutoff values, and a plot with the fit, as well
#' as information as to which polynomial fit the data best.
#'
#' @param data Data frame with the calibration (empty chamber) rapid A/Ci response
#' @param mincut Minimum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the initial chamber mixing. Default value is set to the
#' minimum COR_r value.
#' @param maxcut Maximum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the end of the response. Not needed in all cases. Default
#' value is set to the maximum COR_r value.
#' @param varnames Variable names - this allows for the use of this code with
#' other machines and setups where variable names may differ.
#'
#' @return racircalcheck allows visual checking of RACiR calibration data
#' @importFrom utils read.delim
#' @importFrom stats BIC
#' @importFrom stats lm
#' @importFrom stats predict
#' @importFrom graphics legend
#' @importFrom graphics lines
#' @importFrom graphics plot
#' @export
#' @examples \donttest{
#' #Read in the file
#' data <- read_6800(system.file("extdata", "cal", package = "racir"))
#' #Run calibration check
#' racircalcheck(data = data,
#'               mincut = 350,
#'               maxcut = 780)
#' }
#'
racircalcheck <- function(data, mincut, maxcut,
                          varnames = list(A = "A",
                                          Ca = "Ca",
                                          CO2_r = "CO2_r",
                                          E = "E",
                                          gtc = "gtc")){
  #assign variable names
  data$A <- data[, varnames$A]
  data$Ca <- data[, varnames$Ca]
  data$CO2_r <- data[, varnames$CO2_r]
  data$E <- data[, varnames$E]
  data$gtc <- data[, varnames$gtc]
  # Check data for cutoffs ----------------------------------
  plot(A~CO2_r, data = data, main = "Check cutoffs")
  # Assign cutoffs ------------------------------------------
  mincut <- ifelse(missing(mincut) == TRUE, min(data$CO2_r), mincut)
  maxcut <- ifelse(missing(maxcut) == TRUE, max(data$CO2_r), maxcut)
  data <- data[data$CO2_r > mincut, ]
  data <- data[data$CO2_r < maxcut, ]
  # Confirm cutoffs are appropriate -------------------------
  plot(A~CO2_r, data = data, main = "Confirm cutoffs")
  # Fit polynomials to calibration curve --------------------
  cal1st <- lm(A ~ CO2_r, data = data)
  cal2nd <- lm(A ~ poly(CO2_r, 2), data = data)
  cal3rd <- lm(A ~ poly(CO2_r, 3), data = data)
  cal4th <- lm(A ~ poly(CO2_r, 4), data = data)
  cal5th <- lm(A ~ poly(CO2_r, 5), data = data)
  predict1 <- predict(cal1st)
  predict2 <- predict(cal2nd)
  predict3 <- predict(cal3rd)
  predict4 <- predict(cal4th)
  predict5 <- predict(cal5th)
  # Plot calibration data and curve fits --------------------
  plot(data$A ~ data$CO2_r, main = "Calibration Fits")
  lines(data$CO2_r, predict1, col = "green")
  lines(data$CO2_r, predict2, col = "blue")
  lines(data$CO2_r, predict3, col = "red")
  lines(data$CO2_r, predict4, col = "grey")
  lines(data$CO2_r, predict5, col = "black")
  legend("bottom", horiz = TRUE, title = "Polynomial",
         legend = c("1st", "2nd", "3rd", "4th", "5th"),
         col = c("green", "blue", "red", "grey", "black"),
         lty = 1, cex = 0.6)

  # Use BIC to assess best polynomial -----------------------
  bics <- BIC(cal1st, cal2nd, cal3rd, cal4th, cal5th)
  best <- noquote(rownames(bics)[bics$BIC == min(bics$BIC)])
  print(best)
}


R/racircal_advanced.R
#' Corrects rapid A/Ci response (RACiR) data from leaves using empty chamber data.
#'
#' \code{racircal_advanced} Interval correction for RACiR data.
#'
#' @inheritParams racircal
#' @param digits Specifies rounding for groups. Defaults to -2 (100s). Effectively
#' uses 100 ppm intervals (e.g. data matching >50 ppm to 150 ppm would be assigned
#' to an interval centered around 100 ppm for reference CO2).
#'
#' @return racircal_advanced racircalcheck allows visual checking of RACiR
#' calibration data
#' @importFrom stats BIC
#' @importFrom stats lm
#' @importFrom stats predict
#' @importFrom graphics legend
#' @importFrom graphics lines
#' @importFrom graphics plot
#' @export
#' @examples \donttest{
#' #Read in data
#' data <- read_6800(system.file("extdata", "poplar_2", package = "racir"))
#' caldata <- read_6800(system.file("extdata", "cal", package = "racir"))
#' #Correct data
#' data_corrected <- racircal_advanced(data = data, caldata = caldata,
#'                                     mincut = 350, maxcut = 780,
#'                                     digits = -2, title = "Test")
#' }
#'
racircal_advanced <- function(data, caldata, mincut, maxcut, title,
                              digits, varnames = list(A = "A",
                                                      Ca = "Ca",
                                                      CO2_r = "CO2_r",
                                                      E = "E",
                                                      gtc = "gtc")){
  #assign variable names
  data$A <- data[, varnames$A]
  data$Ca <- data[, varnames$Ca]
  data$CO2_r <- data[, varnames$CO2_r]
  data$E <- data[, varnames$E]
  data$gtc <- data[, varnames$gtc]

  digits <- ifelse(missing(digits) == TRUE, -2, digits)
  # Check for title
  title <- ifelse(missing(title) == TRUE, NA, title)
  #Add separation column --------------------------------------
  caldata$sep_columns <- round(caldata$CO2_r, digits = digits)
  data$sep_columns <- round(data$CO2_r, digits = digits)
  # Assign cutoffs --------------------------------------------
  mincut <- ifelse(missing(mincut) == TRUE, min(caldata$CO2_r), mincut)
  maxcut <- ifelse(missing(maxcut) == TRUE, max(caldata$CO2_r), maxcut)
  caldata <- caldata[caldata$CO2_r > mincut, ]
  caldata <- caldata[caldata$CO2_r < maxcut, ]
  # Assigns maximum and minimum CO2_r values based on ---------
  # calibration data ------------------------------------------
  maxcal <- max(caldata$CO2_r)
  mincal <- min(caldata$CO2_r)
  data <- data[data$CO2_r > mincal, ]
  data <- data[data$CO2_r < maxcal, ]
  #Split data based on CO2 ranges -----------------------------
  range.split_cal <- split(caldata, caldata$sep_columns)
  range.split_data <- split(data, data$sep_columns)

  #Check for overlap between data and calibration file
  intervals <- names(range.split_data) %in% names(range.split_cal)
  int_position <- 1:length(intervals)
  intervals <- data.frame(intervals, int_position)
  for(i in 1:nrow(intervals)){
    if(intervals$intervals[i] == FALSE){
      intervals$intervals[i] <- NA
    }
  }
  intervals <- intervals[is.na(intervals$intervals) == FALSE,]
  range.split_cal <- range.split_cal[intervals$int_position]
  range.split_data <- range.split_data[intervals$int_position]

  for(i in seq_along(range.split_data)){
      if(nrow(range.split_cal[[i]]) < 6 |
         nrow(range.split_data[[i]]) < 6 ){
        error_message <- paste("Interval",
        names(range.split_data)[i],
"ppm has insufficient observations (<6) for the
advanced correction to work. Please adjust mincut or maxcut.")
        stop(error_message)
      }
  }

  #Model calibration and correct leaf data --------------------
  for(i in seq_along(range.split_data)){
    cal1st <- lm(A ~ CO2_r, data = range.split_cal[[i]])
    cal2nd <- lm(A ~ poly(CO2_r, 2), data = range.split_cal[[i]])
    cal3rd <- lm(A ~ poly(CO2_r, 3), data = range.split_cal[[i]])
    cal4th <- lm(A ~ poly(CO2_r, 4), data = range.split_cal[[i]])
    cal5th <- lm(A ~ poly(CO2_r, 5), data = range.split_cal[[i]])
    # Use BIC to assess best polynomial -----------------------
    bics <- BIC(cal1st, cal2nd, cal3rd, cal4th, cal5th)
    best <- noquote(rownames(bics)[bics$BIC == min(bics$BIC)])
    # Correct leaf data ---------------------------------------
    ifelse(best == "cal5th", range.split_data[[i]]$Acor <-
            range.split_data[[i]]$A - predict(cal5th, range.split_data[[i]]),
      ifelse(best == "cal4th", range.split_data[[i]]$Acor <-
            range.split_data[[i]]$A - predict(cal4th, range.split_data[[i]]),
      ifelse(best == "cal3rd", range.split_data[[i]]$Acor <-
            range.split_data[[i]]$A - predict(cal3rd, range.split_data[[i]]),
      ifelse(best == "cal2nd", range.split_data[[i]]$Acor <-
            range.split_data[[i]]$A - predict(cal2nd, range.split_data[[i]]),
        range.split_data[[i]]$Acor <-
          range.split_data[[i]]$A - predict(cal1st, range.split_data[[i]])))))
    range.split_data[[i]]$Cicor = ( ( (range.split_data[[i]]$gtc -
                     range.split_data[[i]]$E / 2) *
                     range.split_data[[i]]$Ca - range.split_data[[i]]$Acor) /
                    (range.split_data[[i]]$gtc + range.split_data[[i]]$E / 2))
  }
  #Recombine data -------------------------------------------
  data <- do.call("rbind", range.split_data)
  # Plot corrected leaf data --------------------------------
  plot(Acor ~ Cicor, data = data, main = title)
  # Plot A vs Cs as well
  plot(Acor ~ CO2_s, data = data, main = title)
  # Remove columns filled with NA ---------------------------
  output <- data[, unlist(lapply(data, function(x) !all(is.na(x))))]
  # Return output
  return(output)
}


R/racircalcheck_advanced.R
#' Allows visual checking of rapid A/Ci response (RACiR) calibration data using
#' empty chamber data.
#'
#' \code{racircalcheck_advanced} Used to check range of calibration file.
#' Produces diagnostic graphs of A vs. Ci for quality control.
#'
#' @inheritParams racircalcheck
#' @param digits Specifies rounding for groups. Defaults to -2 (100s). Effectively
#' uses 100 ppm intervals (e.g. data matching >50 ppm to 150 ppm would be assigned
#' to an interval centered around 100 ppm for reference CO2).
#'
#' @return racircalcheck_advanced returns a data frame with corrected RACiR data
#' @importFrom stats BIC
#' @importFrom stats lm
#' @importFrom stats predict
#' @importFrom graphics legend
#' @importFrom graphics lines
#' @importFrom graphics plot
#' @export
#' @examples \donttest{
#' #Read in data
#' data <- read_6800(system.file("extdata", "poplar_2", package = "racir"))
#' caldata <- read_6800(system.file("extdata", "cal", package = "racir"))
#' #Correct data
#' racircalcheck_advanced(data = data, mincut = 350, maxcut = 780)
#' }
#'
#'
racircalcheck_advanced <- function(data, mincut, maxcut,
                                   digits,
                                   varnames = list(A = "A",
                                                   Ca = "Ca",
                                                   CO2_r = "CO2_r",
                                                   E = "E",
                                                   gtc = "gtc")){
  #assign variable names
  data$A <- data[, varnames$A]
  data$Ca <- data[, varnames$Ca]
  data$CO2_r <- data[, varnames$CO2_r]
  data$E <- data[, varnames$E]
  data$gtc <- data[, varnames$gtc]
  digits <- ifelse(missing(digits) == TRUE, -2, digits)
  #Create intervals
  data$sep_columns <- round(data$CO2_r, digits = digits)

  # Assign cutoffs ------------------------------------------
  mincut <- ifelse(missing(mincut) == TRUE, min(data$CO2_r), mincut)
  maxcut <- ifelse(missing(maxcut) == TRUE, max(data$CO2_r), maxcut)
  data <- data[data$CO2_r > mincut, ]
  data <- data[data$CO2_r < maxcut, ]

  range.split <- split(data, data$sep_columns)
  for(i in seq_along(range.split)){
    if(nrow(range.split[[i]]) < 6){
      error_message <- paste("Interval",
                             names(range.split)[i],
                             "ppm has insufficient observations (<6) for the
advanced correction to work. Please adjust mincut or maxcut.")
      stop(error_message)
    }
  }
  for (i in seq_along(range.split)){
  # Fit polynomials to calibration curve --------------------
  cal1st <- lm(A ~ CO2_r, data = range.split[[i]])
  cal2nd <- lm(A ~ poly(CO2_r, 2), data = range.split[[i]])
  cal3rd <- lm(A ~ poly(CO2_r, 3), data = range.split[[i]])
  cal4th <- lm(A ~ poly(CO2_r, 4), data = range.split[[i]])
  cal5th <- lm(A ~ poly(CO2_r, 5), data = range.split[[i]])
  range.split[[i]]$predict1 <- predict(cal1st)
  range.split[[i]]$predict2 <- predict(cal2nd)
  range.split[[i]]$predict3 <- predict(cal3rd)
  range.split[[i]]$predict4 <- predict(cal4th)
  range.split[[i]]$predict5 <- predict(cal5th)
  # Use BIC to assess best polynomial -----------------------
  bics <- BIC(cal1st, cal2nd, cal3rd, cal4th, cal5th)
  best <- noquote(rownames(bics)[bics$BIC == min(bics$BIC)])
  print(best)
  }
  data <- do.call("rbind", range.split)
  # Plot calibration data and curve fits --------------------
  plot(data$A ~ data$CO2_r, main = "Calibration Fits")
  lines(data$CO2_r, data$predict1, col = "green")
  lines(data$CO2_r, data$predict2, col = "blue")
  lines(data$CO2_r, data$predict3, col = "red")
  lines(data$CO2_r, data$predict4, col = "grey")
  lines(data$CO2_r, data$predict5, col = "black")
  legend("bottom", horiz = TRUE, title = "Polynomial",
         legend = c("1st", "2nd", "3rd", "4th", "5th"),
         col = c("green", "blue", "red", "grey", "black"),
         lty = 1, cex = 0.6)
  }


R/racircalbatch_advanced.R
#' Corrects a batch of rapid A/Ci response (RACiR) data
#'
#' \code{racircalbatch_advanced} Corrects your RACiR data files based on a
#' calibration file. Produces diagnostic graphs of A vs. Ci for quality
#' control. Output includes a list of data frames with corrected data.
#'
#' @param data List of data frames with the RACiR response data
#' @param caldata Data frame with the calibration data
#' @param mincut Minimum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the initial chamber mixing. Default value is set to the
#' minimum COR_r value.
#' @param maxcut Maximum cutoff value for reference CO2 (CO2_r). Used to cut
#' out the data from the end of the response. Not needed in all cases. Default
#' value is set to the maximum COR_r value.
#' @param title Vector for titles of output graph - useful for batch RACiR
#' corrections. Length must be equal to data list length
#' @param varnames Variable names - this allows for the use of this code with
#' other machines and setups where variable names may differ.
#' @param digits Specifies rounding for groups. Defaults to -2 (100s). Effectively
#' uses 100 ppm intervals (e.g. data matching >50 ppm to 150 ppm would be assigned
#' to an interval centered around 100 ppm for reference CO2).
#'
#' @return racircalbatch_advanced uses racircal_advanced on many files
#' @importFrom stats BIC
#' @importFrom stats lm
#' @importFrom stats predict
#' @importFrom graphics legend
#' @importFrom graphics lines
#' @importFrom graphics plot
#' @export
#' @examples \donttest{
#'#Create a list of files
#'files <- c(system.file("extdata", "poplar_1", package = "racir"),
#'           system.file("extdata", "poplar_2", package = "racir"))
#'data <- vector("list", length(files))
#'for(i in seq_along(files)){
#'  data[[i]] <- read_6800(files[i])
#'  names(data)[i] <- files[i]
#'}
#'
#'caldata <- read_6800(system.file("extdata", "cal", package = "racir"))
#'output <- racircalbatch_advanced(caldata = caldata, data = data,
#'                                 mincut = 300, maxcut = 780, title = files)
#'}
racircalbatch_advanced <- function(caldata, data, mincut, maxcut,
                                   digits, title,
                                   varnames = list(A = "A",
                                                   Ca = "Ca",
                                                   CO2_r = "CO2_r",
                                                   E = "E",
                                                   gtc = "gtc")){
  digits <- ifelse(missing(digits) == TRUE, -2, digits)
  # Check for title
  title <- ifelse(missing(title) == TRUE, rep(NA, length(data)), title)
  output <- vector("list", length(data))
  for (i in seq_along(data)){
    output[[i]] <- racircal_advanced(caldata = caldata,
                   data = data[[i]],
                   mincut = mincut,
                   maxcut = maxcut,
                   title = title[i],
                   digits = digits,
                   varnames = varnames)
    output[[i]]$ID <- title[i]
  }
  return(output)
  }
